File: lecture notes/matlab/lec10/run1.m
clear;clc;
T=0.2;
ts=1;
os=10/100;
zeta=-log(os)/sqrt(pi^2+log(os)^2);
wn=4/(ts*zeta);
zpoles=exp(T*roots([1 2*zeta*wn wn^2]));
coef=conv([1 -zpoles(1)],[1 -zpoles(2)]);
Gs=tf(1,[1 2]);
Gz=c2d(Gs,T,'zoh');
syms s z;
Gzz=0.1648/(z-0.6703);


syms kp ki kd real;
Fzz=((kp+ki+kd)*z^2-(kp+2*kd)*z+kd)/(z^2-z);
Tzz=(Fzz*Gzz)/(Fzz*Gzz+1);
[pzs,pcs]=numden(Tzz);
coef_pcs=coeffs(pcs,z,'all');
coef_pcs=coef_pcs/coef_pcs(1);
syms p real;
prob=coef_pcs==coeffs((coef(1)*s^2+coef(2)*s+coef(3))*(s+p),s,'all');
sol=solve(prob,[p kd ki]);

pval=sol.p;
kpmin=double(solve(pval==1));
kpmax=double(solve(pval==-1));

fun_p=matlabFunction(sol.p);
fun_kd=matlabFunction(sol.kd);
fun_ki=matlabFunction(sol.ki);

lenn=400;
kpvec=linspace(kpmin+0.1,kpmax-0.1,lenn);
tablo=zeros(lenn,6);
for i=1:lenn
    kpval=kpvec(i);
    kdval=fun_kd(kpval);
    kival=fun_ki(kpval);
    Fz=tf([kpval+kival+kdval -(kpval+2*kdval) kdval],[1 -1 0],T);
    Tz=feedback(Fz*Gz,1);
    info=stepinfo(Tz); 
    tsval=info.SettlingTime;
    osval=info.Overshoot/100;
    err=0.5*abs(ts-tsval)/ts+0.5*abs(os-osval)/os;
    tablo(i,:)=[kpval kival kdval tsval osval err];
end

[val,index]=min(tablo(:,end));
row=tablo(index,:);
kpval=row(1);
kival=row(2);
kdval=row(3);
tsval=row(4);
osval=row(5);
Fz=tf([kpval+kival+kdval -(kpval+2*kdval) kdval],[1 -1 0],T);
Tz=feedback(Fz*Gz,1);
Tz
zpk(Tz)
stepinfo(Tz)

[y,t]=step(Tz);
figure(1);clf;hold on;grid minor;set(gca, 'MinorGridColor', 'k','MinorGridAlpha',1);
stem(t,y,'k','LineWidth',2);
print("../../img/"+"lec10_step1.eps",'-deps','-r150');

File: lecture notes/python/lec10/run1.py
import matplotlib.pyplot as plt
import numpy as np
import sympy as sym
import tbcontrol
import control
############################
# q1
############################
def fun_q1():
    T=0.2
    p,z,ki,kp,kd=sym.symbols("p,z,ki,kp,kd")
    num=0.1648
    den=z-0.6703
    Gz=control.tf(num,np.array(sym.Poly(den,z).all_coeffs(),dtype=np.float64),T)

    os=10/100
    ts=1
    zeta=-np.log(os)/np.sqrt(np.pi**2+np.log(os)**2)
    wn=4/(ts*zeta)
    
    pole_s=-zeta*wn+1j*wn*np.sqrt(1-zeta**2)
    pole_z=np.exp(pole_s*T)
    pdz=(z-pole_z)*(z-np.conj(pole_z))*(z-p)
    pdz=sym.expand(pdz)

    Gzz=num/den
    Fzz=((kp+ki+kd)*z**2-(kp+2*kd)*z+kd)/(z**2-z)

    Tzz=(Fzz*Gzz)/(1+Fzz*Gzz)
    Tzz=sym.simplify(Tzz)
    pzz,pcz=sym.fraction(Tzz)
    pcz=sym.expand(pcz)

    coef_pcz=sym.Poly(pcz,z).all_coeffs()
    coef_pdz=sym.Poly(pdz,z).all_coeffs()
    prob=[]
    for i in range(1,len(coef_pcz)):
        e1=coef_pdz[i]/coef_pdz[0]
        e2=coef_pcz[i]/coef_pcz[0]
        prob.append(e1-e2)
    sol=sym.solve(prob)
    kdval=sol[kd]
    kival=sol[ki]
    kpval=sol[kp]

    pvec=np.arange(0.1,1,0.1)
    cost=np.zeros(pvec.shape)
    for i in range(0,len(pvec)):
        pval=pvec[i]
        kdv=float(kdval.subs(p,pval))
        kiv=float(kival.subs(p,pval))
        kpv=float(kpval.subs(p,pval))
        Fz=control.tf(np.array([kdv+kiv+kpv,-kpv-2*kdv,kdv]),np.array([1,-1,0]),T)
        Tz=control.feedback(control.series(Fz,Gz),1)
        info=control.step_info(Tz)
        tsval=info['SettlingTime']
        osval=info['Overshoot']/100
        print(osval)
        cost[i]=0.5*np.abs(tsval-ts)/ts+0.5*np.abs(osval-os)/os
    
    index=np.argmin(cost)
    pval=pvec[index]
    kdv=float(kdval.subs(p,pval))
    kiv=float(kival.subs(p,pval))
    kpv=float(kpval.subs(p,pval))
    Fz=control.tf(np.array([kdv+kiv+kpv,-kpv-2*kdv,kdv]),np.array([1,-1,0]),T)
    Tz=control.feedback(control.series(Fz,Gz),1)
    t, y = control.step_response(Tz)
    info=control.step_info(Tz)
    
    print(Fz)
    print(Tz)
    plt.stem(t,y)
    plt.xlim([0,4])
    plt.title("Basamak Yanıtı")
    plt.xlabel("Zaman(s)")
    plt.show()
    
fun_q1()